spec/services/fetch_coronavirus_statistics_service_spec.rb

```diff
@@ -0,0 +1,124 @@
+          { "date" => "2021-03-17", "cumulativeVaccinations" => 25_735_472, "hospitalAdmissions" => nil, "newPositiveests" => 5758 },
+          { "date" => "2021-03-16", "cumulativeVaccinations" => 25_273_226, "hospitalAdmissions" => nil, "newPositiveests" => 5294 },
+          { "date" => "2021-03-15", "cumulativeVaccinations" => 24_839_906, "hospitalAdmissions" => nil, "newPositiveests" => 5089 },
+          { "date" => "2021-03-14", "cumulativeVaccinations" => 24_453_221, "hospitalAdmissions" => 426, "newPositiveests" => 4618 },
+          { "date" => "2021-03-13", "cumulativeVaccinations" => 24_196_211, "hospitalAdmissions" => 460, "newPositiveests" => 5534 },
+        ],
+      }
+
+      stub_request(:get, /coronavirus.data.gov.uk/)
+        .to_return(status: 200, body: body.to_json)
+
+      statistics = described_class.call
+      expect(statistics).to be_a(described_class::Statistics)
+      expect(statistics).to have_attributes(
+        cumulative_vaccinations: 25_735_472,
+        cumulative_vaccinations_date: Date.new(2021, 3, 17),
+        hospital_admissions: 426,
+        hospital_admissions_date: Date.new(2021, 3, 14),
+        new_positive_tests: 6303,
+        new_positive_tests_date: Date.new(2021, 3, 18),
+      )
+    end
+
+    context "when there is missing data" do
+      it "returns nil for no data" do
+        stub_request(:get, /coronavirus.data.gov.uk/)
+          .to_return(status: 200, body: { data: [] }.to_json)
+
+        expect(described_class.call).to be_nil
+      end
+
+      it "sets only the fields that have data" do
+        body = { data: [{ "date" => "2021-03-18",
+                          "cumulativeVaccinations" => nil,
+                          "hospitalAdmissions" => nil,
+                          "newPositiveTests" => 6303 }] }
+
+        stub_request(:get, /coronavirus.data.gov.uk/)
+          .to_return(status: 200, body: body.to_json)
+
+        expect(described_class.call).to have_attributes(
+          cumulative_vaccinations: nil,
+          cumulative_vaccinations_date: nil,
+          hospital_admissions: nil,
+          hospital_admissions_date: nil,
+          new_positive_tests: 6303,
+          new_positive_tests_date: Date.new(2021, 3, 18),
+        )
+      end
+    end
+
+    context "when the request to load data fails" do
+      it "notifies GovukError for a general error" do
+        stub_request(:get, /coronavirus.data.gov.uk/).to_return(status: 500)
+        expect(GovukError).to receive(:notify)
+
+        described_class.call
+      end
+
+      it "doens't notify GovukError when a timeout occurs" do
+        stub_request(:get, /coronavirus.data.gov.uk/).to_timeout
+        expect(GovukError).not_to receive(:notify)
+
+        described_class.call
+      end
+
+      it "doens't notify GovukError on 502 and 504 errors" do
+        stub_request(:get, /coronavirus.data.gov.uk/)
+          .to_return(status: 502)
+          .then
```

Comment:
This is nice! I didn't know it was possible to do this.

Reply:
I had to Google 😁